### PR TITLE
Issue/29 add image resource

### PR DESCRIPTION
**Pull request by kazaamjt**

### CODE REVIEW

File: plugins/__init__.py [hunk 1]
```diff
@@ -865,6 +885,141 @@ def update_resource(self, ctx: handler.HandlerContext, changes: dict, resource:
...
+                image = self._glance.images.get(image.id)
+                if image.status == "active":
+                    break
+                time.sleep(0.1)
+
+        ctx.set_created()
+
+    def delete_resource(self, ctx: handler.HandlerContext, resource: resources.PurgeableResource):
+        image = self._get_image(resource)
+        self._glance.images.delete(image.id)
```
Comment (arnaudsjs) [inline plugins/__init__.py]: Use the image_id from the ctx. This saves an API call.

File: plugins/__init__.py [hunk 2]
```diff
@@ -865,6 +885,141 @@ def update_resource(self, ctx: handler.HandlerContext, changes: dict, resource:
...
+
+        ctx.set_purged()
+
+    def update_resource(self, ctx: handler.HandlerContext, changes: dict, resource: resources.PurgeableResource):
```
Comment (arnaudsjs) [inline plugins/__init__.py]: This method does an API call per attribute. Is it possible to reduce this to one API call?

File: model/_init.cf [hunk 3]
```diff
@@ -465,3 +466,41 @@ entity Flavor extends OpenStackResource:
...
+        :attr visibility: Can be set to public or private. Shared and Community are currently not implemented.
+        :attr protected: Wether the image can be deleted or not. Inmanta will never delete protected images.
+        :attr metadata: Various metadata passed as a dict.
+        :attr skip_on_deploy: When set, inmanta will not wait for the image to be deployed and mark it as skipped.
+        :attr purge_on_delete: When set to true, the image will be removed when no longer present in the model.
+    """
+    string name
+    string uri
+    container_format? container_format = "bare"
+    disk_format? disk_format = "qcow2"
```
Comment (arnaudsjs) [inline model/_init.cf]: Are these fields optional?

Comment (kazaamjt) [reply model/_init.cf]: They specifically can be set to None in the OpenStack API.

File: plugins/__init__.py (hunk 2 above)
Comment (kazaamjt) [reply plugins/__init__.py]: Create a dictionary with all the data and unpack it?
That would reduce it to 1 API call.

File: model/_init.cf [hunk 4]
```diff
@@ -465,3 +466,41 @@ entity Flavor extends OpenStackResource:
 end
 
 implement Flavor using providerRequire
+
+index Image (provider, name)
+Image.provider [1] -- Provider.images [0:]
+
+typedef container_format as string matching self in ["ami", "ari", "aki", "bare", "ovf", "ova", "docker"]
+typedef disk_format as string matching self in ["ami", "ari", "aki", "vhd", "vhdx", "vmdk", "raw", "qcow2", "vdi", "iso", "ploop"]
+# TODO: add support for "shared" and "community"
```
[the remark this answers is not in the document]
Comment (kazaamjt) [reply model/_init.cf]: resolved

File: model/_init.cf [hunk 5]
```diff
@@ -465,3 +466,41 @@ entity Flavor extends OpenStackResource:
...
+entity Image extends OpenStackResource:
+    """
+        A machine image for OpenStack VMs
+
+        :attr name: Name for the flavor. Inmanta treats images names as unique per provider.
+        :attr uri: a link to the download location of the image.
+        :attr container_format: Must be one of [null, ami, ari, aki, bare, ovf, ova, docker].
+        :attr disk_format: Must be one of [null, ami, ari, aki, vhd, vhdx, vmdk, raw, qcow2, vdi, iso, ploop].
+        :attr image_id: uuid to identify the image. Auto set by OpenStack if not set.
+        :attr visibility: Can be set to public or private. Shared and Community are currently not implemented.
```
[the remark this answers is not in the document]
Comment (kazaamjt) [reply model/_init.cf]: Resolved

File: plugins/__init__.py (hunk 1 above)
Comment (kazaamjt) [reply plugins/__init__.py]: Resolved.
Also did this for update resource.

File: plugins/__init__.py (hunk 2 above)
Comment (kazaamjt) [reply plugins/__init__.py]: Resolved, implemented mentioned change

File: model/_init.cf (hunk 3 above)
Comment (arnaudsjs) [reply model/_init.cf]: Ok.

File: model/_init.cf [hunk 6]
```diff
@@ -465,3 +466,41 @@ entity Flavor extends OpenStackResource:
...
+entity Image extends OpenStackResource:
+    """
+        A machine image for OpenStack VMs
+
+        :attr name: Name for the flavor. Inmanta treats image names as unique per provider.
+        :attr uri: a link to the download location of the image.
+        :attr container_format: Must be one of [null, ami, ari, aki, bare, ovf, ova, docker].
+        :attr disk_format: Must be one of [null, ami, ari, aki, vhd, vhdx, vmdk, raw, qcow2, vdi, iso, ploop].
+        :attr image_id: uuid to identify the image. Auto set by OpenStack if not set.
+        :attr visibility: Wether the image is visible across all projects. Can either be public or private. Shared and community are currently not implemented.
```
Comment (arnaudsjs) [inline model/_init.cf]:
```suggestion
        :attr visibility: Whether the image is visible across all projects. Can either be public or private. Shared and community are currently not implemented.
```

File: plugins/__init__.py [hunk 7]
```diff
@@ -865,6 +885,150 @@ def update_resource(self, ctx: handler.HandlerContext, changes: dict, resource:
...
+                image = self._glance.images.get(image.id)
+                if image.status == "active":
+                    break
+                time.sleep(0.1)
+
+        ctx.set_created()
+
+    def delete_resource(self, ctx: handler.HandlerContext, resource: resources.PurgeableResource):
+        image = self._glance.images.get(ctx.get("image_id"))
+        self._glance.images.delete(image.id)
```
Comment (arnaudsjs) [inline plugins/__init__.py]: This code does two API calls: one to get the image and one to delete the image:
Would be better to use: `self._glance.images.delete(ctx.get("image_id"))`

Comment (kazaamjt) [reply plugins/__init__.py]: resolved

**Pull request merged**